src/middlewares/alert.ts

```diff
@@ -0,0 +1,58 @@
+import { create as createLogger } from '../common/log'
+const log = createLogger('alert-middleware')
+import { Middleware, MiddlewareCallback, Pipelines } from '../types/middleware'
+import * as IlpPacket from 'ilp-packet'
+
+const { T04_INSUFFICIENT_LIQUIDITY } = IlpPacket.Errors.codes
+
+export interface Alert {
+  id: number
+  accountId: string
+  createdAt: Date
+  message: string
+}
+
+export default class AlertMiddleware implements Middleware {
+  private alerts: {[id: string]: Alert} = {}
+  private nextAlertId: number = Date.now()
+
+  async applyToPipelines (pipelines: Pipelines, accountId: string) {
+    pipelines.outgoingData.insertLast({
+      name: 'alert',
+      method: async (data: Buffer, next: MiddlewareCallback<Buffer, Buffer>) => {
+        const result = await next(data)
+        if (result[0] !== IlpPacket.Type.TYPE_ILP_REJECT) return result
+
+        const rejectPacket = IlpPacket.deserializeIlpReject(result)
+        if (rejectPacket.code !== T04_INSUFFICIENT_LIQUIDITY) return result
+
+        // The peer rejected a packet which, according to the local balance, should
+        // have succeeded. This can happen when our local connector owes the peer
+        // money but restarted before it was settled.
+        if (rejectPacket.message !== 'exceeded maximum balance.') return result
+
+        log.warn('generating alert for account=%s message="%s"', accountId, rejectPacket.message)
+
+        const id = this.nextAlertId++
+        this.alerts[id] = {
```

Comment:
What do you think of a `count` field on the alert? If an alert occurs, it will check if one with the same peer/message combination already exists, and if so, increment the count instead of creating a new alert.

Reply:
that sounds great, yeah